### ISSUE TITLE
RotatedRect::boundingRect() returns an int rectangle

### ISSUE THREAD

**Issue by PhilLab** [opencv/opencv#6219]
### Please state the information for your system
- OpenCV version: 3.0
- Host OS: all
### In which part of the OpenCV library you got the issue?
- core
### Expected behaviour

`RotatedRect::boundingRect()` should return a `cv::Rect2f` since `cv::RotatedRect` contains floating point data
### Actual behaviour

A rounded integer rectangle is returned [(source)](https://github.com/Itseez/opencv/blob/56c89e0a980d39d8ee41752e896f08a267396873/modules/core/src/matrix.cpp#L5510-L5521), which is also not the minimal one (see code example at the bottom)
### Additional description

The [documentation](http://docs.opencv.org/3.1.0/db/dd6/classcv_1_1RotatedRect.html#a8f44c5ef4f867965c953a721d8a033d9) states 

> returns the minimal up-right rectangle containing the rotated rectangle

and does not mention the precision truncation. When using RotatedRect in [0,1]-coordinate systems, the result is not usable since it always is the rectangle (0,0) -> (2,2). (BTW: shouldn't it be (0,0) -> (1,1)? But it isn't )
### Code example to reproduce the issue / Steps to reproduce the issue

```
#include "opencv2/core.hpp"
#include <iostream>

int main()
{
        cv::RotatedRect rect(cv::Point2f(0.5f, 0.5f), cv::Size2f(0.1f, 0.1f), 0.f);
        std::cout << rect.center << "|" << rect.size << "|" << rect.angle << std::endl;
        std::cout << rect.boundingRect().tl() << "|" << rect.boundingRect().br();
}
```

Console output:

> [0.5, 0.5] | [0.1 x 0.1] | 0
> [0, 0] | [2, 2]

Expected output: 

> [0.5, 0.5] | [0.1 x 0.1] | 0
> [0.45, 0.45] | [0.55, 0.55]

**Comment by alalek** [comment]
1) I believe integer-based return value type is "by design" to use it with images. We should not change current behavior, but you can suggest new method "boundingRect_f"
2) `(2,2)` looks very strange. For unknown reason there is [added](https://github.com/Itseez/opencv/blob/56c89e0a980d39d8ee41752e896f08a267396873/modules/core/src/matrix.cpp#L5518-L5519) 1 to width and height.

**Comment by PhilLab** [comment]
1) okay, I would propose to call it boundingRect2f to be consistent with the typedef. But we have a problem:

2) Maybe the boundingRect definition is defined with the corners exclusively, i.e. (2,2) is the lower right "coordinate" of the lower right pixel. And there might be implementations out there counting on that fact :-/

But if 2) is correct, what should boundingRect2f return? There is no next bigger number (well, technically there is for a float but mathematically there is not). If we would define the bounding box to be inclusive here, this would not be consistent with the original boundingRect method.

**Comment by FredFG** [comment]
One simple approach which doesn't need a new OpenCv method is to "magnify" the rectangle by a factor of  say 1000, rotate it and then divide the resulting dimensions by 1000.0 (float). Obviously by changing the multiplier you can achieve any desired level of precision. Now you have both the rotated, pixellated rectangle and the floating point rectangle with the precision you desire.

pseudocode:

Rect pixellatedBounds = boundingRect(pointsArray); // guaranteed to fall on pixel boundaries.
Rect bigBounds = pixellatedBounds * 1000; // magnified and pixellated Rect, still on pixel boundaries.
Rect floatBounds = bigBounds /1000.0; // floating point bounds which you can use for all sorts of geometric computations.

An ancient trick.

**Comment by bedbad** [comment]
Let's assume it is all integers.
(2,2) is incorrect. It is not exclusively defined for smaller corner, so , for square (0,0) (1,1) it will be partially covering bigger square (0,0) (2, 2) which is no rational sense correct.
I propose pull request for this.
And there should also be boundingRect_f() function, which returns precise number without any intends for drawing images.

**Comment by alalek** [comment]
By design cv::Rect (integer) pixels are:
```
rect.x <= x < rect.x + rect.width
rect.y <= y < recx.y + rect.height
```
(top-left side is inclusive, bottom-right size is exclusive).

So there is nothing to fix with Rect here.

P.S. I believe `boundingRect2f` should return corners with "inclusive" coordinates.

**Comment by bedbad** [comment]
Is there any place where it is documented? I just couldn't find it. 
the fact that brc is tlc + full size can also mean that the size itself should be "inclusive"

P.S: for boundingRect2f it is clearly only "inclusive" because it's supposed to represent numbers for which there is no monotone iteration + there is no practical way to do it

**Comment by alalek** [comment]
> documented

http://docs.opencv.org/master/d2/d44/classcv_1_1Rect__.html#details

**Comment by bedbad** [comment]
thanks for the link. It seemed strange that yourself didn't recognize it in March
What should boundingRect2f return? Point2f, Rect_<float> or should it be template in your opinion?

**Comment by bedbad** [comment]
this code shows the difference between boundingRect2f() and boundingRect()
`
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <iostream>
#include <algorithm>

using namespace cv;

int main(int ac, char ** argv)
{

        Mat image(250, 250, CV_8UC3, Scalar(0, 0, 0));
	RotatedRect rRect = RotatedRect(Point2f(125.0,125.0), Size2f(40.0,40.0), 0);
	Point2f vs[4];
	rRect.points(vs);
                                                                                               
        Rect_<float> brect = rRect.boundingRect2f();
        Rect rect = rRect.boundingRect();

//--------------------------------------------------by lines------------------------------------------------------
        for (int i = 0; i < 4; i++)
	  line(image, vs[i], vs[(i+1)%4], Scalar(0,255,0));
	
	line(image, Point_<int>(rect.x, rect.y), Point_<int>(rect.x+rect.width, rect.y), Scalar(255, 0, 0));
        line(image, Point_<int>(rect.x+rect.width, rect.y), Point_<int>(rect.x+rect.width, rect.y+rect.height), Scalar(255, 0, 0));
        line(image, Point_<int>(rect.x+rect.width, rect.y+rect.height), Point_<int>(rect.x, rect.y+rect.height), Scalar(255, 0, 0));
      	line(image, Point_<int>(rect.x, rect.y+rect.height), Point_<int>(rect.x, rect.y), Scalar(255, 0, 0));

	
	imshow("original vs standard: lines", image);
//----------------------------------------------------------------------------------------------------------
        std::cout << Point_<float>(vs[1].x, vs[1].y) << "|" << Point_<float>(vs[3].x, vs[3].y) << "<--original" << rRect.angle << std::endl;                                                                      
        std::cout << brect.tl() << "|" << brect.br() << "boundingRect2f" << std::endl;
	std::cout << rect.tl() << "|" << rect.br()  << "standard bounding" <<std::endl;
        std::cout << "<-- Rectangles' diff"<<std::endl;

//-------------------------------by rect()---------------------------------------------------------------
	Mat image2(200, 200, CV_8UC3, Scalar(0));
	rectangle(image2, Point_<int>(vs[1].x, vs[1].y), Point_<int>(vs[3].x, vs[3].y), Scalar(0, 255, 0));
	rectangle(image2, rect, Scalar(255, 0, 0));

	imshow("original vs standard: rect", image2);
//--------------------------------by rect() with boundingRect2f------------------------------------------
        Mat image3(200, 200, CV_8UC3, Scalar(0));
        rectangle(image3, Point_<int>(vs[1].x, vs[1].y), Point_<int>(vs[3].x, vs[3].y), Scalar(0, 255, 0));
        rectangle(image3, brect, Scalar(255, 0, 0));

        imshow("2f vs standard: rect()", image3);

	waitKey(0);

}`

**Comment by alalek** [comment]
These keywords works for "default" branch only ("master").
This PR was merged into 2.4 initially (it should be closed manually after "merge-X" PR, but probably this step is missed).
OOM PR is not merged into default branch yet (I will do this today/tomorrow).
